# GENERIC CODE
# Prime Generator

In [ ]:
limit = 10**6
fname = 'primes.csv'

import os.path
if os.path.isfile(fname):
    file = open(fname, 'r')
    buffer = file.read()
    if len(buffer) > 0:
        primes = list(map(int,buffer.split(',')))
    else:
        primes = [2,3]    
    file.close()


    
if len(primes) > 0:
    primecandidates = primes[0:-1] + list(range(primes[-1]+2,limit+1,2))
else:
    primecandidates = list(range(2,limit+1))
primes = []

while len(primecandidates) > 0:
    #old code. too slow
    #primecandidates = [i for i in primecandidates if i not in set(range(primecandidates[0]*2,limit+1,primecandidates[0]))]
    #primes.append(primecandidates.pop(0)) 
    
    #hoping this is faster
    #primes.append(primecandidates.pop(0))
    #primecandidates = [i for i in primecandidates if i%primes[-1] != 0]
    
    primes.append(primecandidates.pop(0))
    nonprimes = list(range(primes[-1]*2,target+1,primes[-1]))     
    primecandidates = [x for x in primecandidates if x not in nonprimes]
 
file = open(fname, 'w')
file.write(','.join(map(str,primes)))
file.close()

print(f'{len(primes)} Prime Numbers Loaded.')
print(f'Highest prime loaded {primes[-1]}')

def isprime(num):
    return num in primes 

def isnotprime(num):
    return num not in primess  

In [26]:
#mnual save on an interrupt
file = open(fname, 'w')
file.write(','.join(map(str,primes)))
file.close()

print(f'{len(primes)} Prime Numbers Loaded.')
print(f'Highest prime loaded {primes[-1]}')

183211 Prime Numbers Loaded.
Highest prime loaded 2502133


## Num_to_List (and List_to_Num)

In [ ]:
def num_to_list(num):
    return [int(letter) for letter in str(num)]

def list_to_num(listinput):
    return int(''.join(map(str,listinput)))

#read the question stupid... not needed
#generates all permutations of a list
def permutations(inputlist): 
    output = []
    for i in inputlist:
        if len(inputlist) == 1:
            return [[i]]
        else:                            
            output = output + list([[i] + j for j in permutations(list(inputlist[0:inputlist.index(i)] + inputlist[inputlist.index(i)+1:]))])         
    return output

#read the question stupid... not needed
#generates all permutations of digits in an integer
def num_perm(num): 
    return list(set(list(map(list_to_num,permutations(num_to_list(num))))))

#generates all rotations of digits in an integer
def num_rotates(num):
    output = []
    for i in range(0,len(str(num))):
        output.append(int(str(num)[i:]+str(num)[:i]))
    return output

# Multiples of 3 and 5
## Problem 1 
If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.

In [1]:
threes = set(range(3,1000,3))

In [3]:
fives = set(range(5,1000,5))

In [6]:
sum(threes | fives)

233168

# Even Fibonacci numbers
## Problem 2 
Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms.

In [26]:
fibonacci = [1,2]
while fibonacci[-2]+fibonacci[-1] < 4 * 10 ** 6:
    fibonacci.append(fibonacci[-2]+fibonacci[-1])
    
fibonacci_even = [f for f in fibonacci if f%2 == 0]
fibonacci_odd = [f for f in fibonacci if f%2 != 0] 

#print(fibonacci)
#print(fibonacci_even)
#print(fibonacci_odd)
print('Sum of even fibonacci numbers below 4million: {}'.format(sum(fibonacci_even)))

Sum of even fibonacci numbers below 4million: 4613732


# Largest prime factor
## Problem 3 
The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ?

In [76]:
number = 600851475143   #WARNING THIS IS TOO SLOW
#number = 100000

#seedprimes
primes = [2]
primefactors = []

#Generate Prime Numbers
tentative_primes = list(range(3,int(number ** 0.5 + 1),2))
while len(tentative_primes) > 0:
    for prime in primes:
        if tentative_primes[0]%prime == 0:
            #not prime
            tentative_primes.pop(0)
            break
        if prime == primes[-1]:
            #is prime
            primes.append(tentative_primes.pop(0))
            break
            
#Prime factorisation
working = number
while len(primes) > 0:
    if working%primes[0] == 0:
        working = working/primes[0]
        primefactors.append(primes[0])
    else:
        primes.pop(0)
        

print('Factorisation : {}'.format('x'.join(map(str,primefactors))))
print('Largest prime factor: {}'.format(primefactors[-1]))

Factorisation : 71x839x1471x6857
Largest prime factor: 6857


# Largest palindrome product
## Problem 4 
A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.

In [116]:
threedigits = list(range(100,1000,1))
palindromes = {}


for n1 in threedigits:
    for n2 in threedigits:
        candidate = str(n1 * n2)
        while candidate[0] == candidate[-1]:
            candidate = candidate[1:-1]
            if len(candidate) <= 1:
                break
                
            
        if len(candidate) <= 1:
            palindromes.update({n1 * n2:(n1,n2)}) 
    
print('Largest palindrome: {} ({}x{})'.format(max(palindromes.keys()),palindromes[max(palindromes.keys())][0],palindromes[max(palindromes.keys())][1]))

Largest palindrome: 906609 (993x913)


# Smallest multiple
## Problem 5 
2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

In [152]:
#seedprimes
primes = [2]

#Generate Prime Numbers
tentative_primes = list(range(3,21,2))
while len(tentative_primes) > 0:
    for prime in primes:
        if tentative_primes[0]%prime == 0:
            #not prime
            tentative_primes.pop(0)
            break
        if prime == primes[-1]:
            #is prime
            primes.append(tentative_primes.pop(0))
            break
            
output = 1
for prime in primes:
    power = 1
    while prime ** (power + 1) <= 20:
        power += 1
    output = output * (prime ** power)
    print('{} {}'.format(prime,power))
    
print('Smallest number: {}'.format(output))

print('\n')
for divisor in range(1,21):
    print(f'{output} divided by {divisor} is {output/divisor}')

2 4
3 2
5 1
7 1
11 1
13 1
17 1
19 1
Smallest number: 232792560


232792560 divided by 1 is 232792560.0
232792560 divided by 2 is 116396280.0
232792560 divided by 3 is 77597520.0
232792560 divided by 4 is 58198140.0
232792560 divided by 5 is 46558512.0
232792560 divided by 6 is 38798760.0
232792560 divided by 7 is 33256080.0
232792560 divided by 8 is 29099070.0
232792560 divided by 9 is 25865840.0
232792560 divided by 10 is 23279256.0
232792560 divided by 11 is 21162960.0
232792560 divided by 12 is 19399380.0
232792560 divided by 13 is 17907120.0
232792560 divided by 14 is 16628040.0
232792560 divided by 15 is 15519504.0
232792560 divided by 16 is 14549535.0
232792560 divided by 17 is 13693680.0
232792560 divided by 18 is 12932920.0
232792560 divided by 19 is 12252240.0
232792560 divided by 20 is 11639628.0


# Sum square difference
## Problem 6 
The sum of the squares of the first ten natural numbers is,

12 + 22 + ... + 102 = 385
The square of the sum of the first ten natural numbers is,

(1 + 2 + ... + 10)2 = 552 = 3025
Hence the difference between the sum of the squares of the first ten natural numbers and the square of the sum is 3025 − 385 = 2640.

Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [169]:
n = 100
sumofsquares = sum(num ** 2 for num in range(1,n+1))
print(sumofsquares)

squareofsum = sum(num for num in range(1,n+1))**2
print(squareofsum)

print(squareofsum - sumofsquares)

338350
25502500
25164150


# 10001st prime
## Problem 7 
By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10001st prime number?

In [170]:
primes = [2]


In [194]:
target = 10001

#Generate Prime Numbers
tentative_prime = primes[-1] + 1
while len(primes) + 1 <= target:
    
    for prime in primes:
        if tentative_prime%prime == 0:
            #not prime
            tentative_prime += 1
            break
        if prime == primes[-1]:
            #is prime
            primes.append(tentative_prime)
            primefound = True
            break

print('Prime no. {} is {}'.format(target,primes[target - 1]))


Prime no. 10001 is 104743


# Largest product in a series
## Problem 8 
The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.

73167176531330624919225119674426574742355349194934
96983520312774506326239578318016984801869478851843
85861560789112949495459501737958331952853208805511
12540698747158523863050715693290963295227443043557
66896648950445244523161731856403098711121722383113
62229893423380308135336276614282806444486645238749
30358907296290491560440772390713810515859307960866
70172427121883998797908792274921901699720888093776
65727333001053367881220235421809751254540594752243
52584907711670556013604839586446706324415722155397
53697817977846174064955149290862569321978468622482
83972241375657056057490261407972968652414535100474
82166370484403199890008895243450658541227588666881
16427171479924442928230863465674813919123162824586
17866458359124566529476545682848912883142607690042
24219022671055626321111109370544217506941658960408
07198403850962455444362981230987879927244284909188
84580156166097919133875499200524063689912560717606
05886116467109405077541002256983155200055935729725
71636269561882670428252483600823257530420752963450

Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?

In [262]:
#Number
numberstr  = '73167176531330624919225119674426574742355349194934'
numberstr += '96983520312774506326239578318016984801869478851843'
numberstr += '85861560789112949495459501737958331952853208805511'
numberstr += '12540698747158523863050715693290963295227443043557'
numberstr += '66896648950445244523161731856403098711121722383113'
numberstr += '62229893423380308135336276614282806444486645238749'
numberstr += '30358907296290491560440772390713810515859307960866'
numberstr += '70172427121883998797908792274921901699720888093776'
numberstr += '65727333001053367881220235421809751254540594752243'
numberstr += '52584907711670556013604839586446706324415722155397'
numberstr += '53697817977846174064955149290862569321978468622482'
numberstr += '83972241375657056057490261407972968652414535100474'
numberstr += '82166370484403199890008895243450658541227588666881'
numberstr += '16427171479924442928230863465674813919123162824586'
numberstr += '17866458359124566529476545682848912883142607690042'
numberstr += '24219022671055626321111109370544217506941658960408'
numberstr += '07198403850962455444362981230987879927244284909188'
numberstr += '84580156166097919133875499200524063689912560717606'
numberstr += '05886116467109405077541002256983155200055935729725'
numberstr += '71636269561882670428252483600823257530420752963450'

width = 13

adjacent = {}
for i in range(0,len(numberstr) - width): #)
    calcbuffer = 1
    for j in range(0, width):
        calcbuffer = calcbuffer * int(numberstr[i + j])  
        #print(int(numberstr[i + j])) 
    
    adjacent.update({i:calcbuffer})
    #print(calcbuffer)
    #print()
    
#print(adjacent)
print('Highest adjacent value product of width {} is {}'.format(width,max(adjacent.values())))
print('First found at poition {} with:'.format([k for k,v in adjacent.items() if v == max(adjacent.values())][0]))
print(numberstr[[k for k,v in adjacent.items() if v == max(adjacent.values())][0]:[k for k,v in adjacent.items() if v == max(adjacent.values())][0]+width])

Highest adjacent value product of width 13 is 23514624000
First found at poition 197 with:
5576689664895


# Special Pythagorean triplet
## Problem 9 
A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

a2 + b2 = c2
For example, 32 + 42 = 9 + 16 = 25 = 52.

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.

In [319]:
length = 1000
triplet = []

for c in range(length,1,-1):
    if a**2 + b**2 == c**2:
        break
    else:    
        for b in range(min(length-c,c-1),1,-1):
            a = length - c - b
            if a > b:
                break
            if a**2 + b**2 == c**2:
                triplet.append((a,b,c,length, a*b*c))
                break
            else:
                continue
                
print('a={}, b={}, c={}, a+b+c={}, abc={}'.format(triplet[0][0],triplet[0][1],triplet[0][2],triplet[0][3],triplet[0][4]))

a=200, b=375, c=425, a+b+c=1000, abc=31875000


# Summation of primes
## Problem 10 
The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.

In [329]:
primes = [2,3]

In [ ]:
#loadprimes - DO THIS IF YOU DO NOT HAVE TWO HOURS
outputfile = open('C:\\Users\\alexp\\primes.csv','r')
primes = outputfile.read()
primes = list(map(int,newprimes.split(',')))
outputfile.close()

In [374]:
target = 2 * 10 ** 6

#Generate Prime Numbers
tentative_prime = primes[-1] + 2
while tentative_prime <= target:
    
    for prime in primes:
        if tentative_prime%prime == 0:
            #not prime
            tentative_prime += 2
            break
        if prime == primes[-1]:
            #is prime
            primes.append(tentative_prime)
            primefound = True
            break

print('There are {} primes below {}. The largest prime is {}. The sum of these primes is {}'.format(len(primes),target,primes[-1],sum(primes)))

There are 148933 primes below 2000000. The largest prime is 1999993. The sum of these primes is 142913828922


In [400]:
#save primes
outputfile = open('C:\\Users\\alexp\\primes.csv','w')
outputfile.write(','.join(map(str,primes)))
outputfile.close()

Alternate prime generation test - Sieve of Eratosthenes
https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes

In [68]:
target =100000
primecandidates = [2] + list(range(3,target+1,2))
n=1
#print(primecandidates)
while n < target and n < len(primecandidates):
    nonprimes = list(range(primecandidates[n]*2,target+1,primecandidates[n])) #list(range(primecandidates(n)*2,target+1,primecandidates(n)))
    
    #primecandidates = list(set(primecandidates) - set(nonprimes))
    #sets are unordered. This was causing problems on the first attempt
    
    primecandidates = [x for x in primecandidates if x not in nonprimes]
    #print(primecandidates)
    n += 1



# Largest product in a grid
## Problem 11 
In the 20×20 grid below, four numbers along a diagonal line have been marked in red.

08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50
32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70
67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21
24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72
21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69
04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36
20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16
20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54
01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48

The product of these numbers is 26 × 63 × 78 × 14 = 1788696.

What is the greatest product of four adjacent numbers in the same direction (up, down, left, right, or diagonally) in the 20×20 grid?

In [123]:
52*49*99*97

24468444

In [139]:
#grid

grid = []
grid.append(list(map(int,'08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08'.split())))
grid.append(list(map(int,'49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00'.split())))
grid.append(list(map(int,'81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65'.split())))
grid.append(list(map(int,'52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91'.split())))
grid.append(list(map(int,'22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80'.split())))
grid.append(list(map(int,'24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50'.split())))
grid.append(list(map(int,'32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70'.split())))
grid.append(list(map(int,'67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21'.split())))
grid.append(list(map(int,'24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72'.split())))
grid.append(list(map(int,'21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95'.split())))
grid.append(list(map(int,'78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92'.split())))
grid.append(list(map(int,'16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57'.split())))
grid.append(list(map(int,'86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58'.split())))
grid.append(list(map(int,'19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40'.split())))
grid.append(list(map(int,'04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66'.split())))
grid.append(list(map(int,'88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69'.split())))
grid.append(list(map(int,'04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36'.split())))
grid.append(list(map(int,'20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16'.split())))
grid.append(list(map(int,'20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54'.split())))
grid.append(list(map(int,'01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48'.split())))

width = 4
results = {}

for i  in range(0 , 21-width):
    for j in range(0 , 21-width):
        
        down = 1
        accross = 1 
        diag_dr = 1
        diag_ur = 1
        for k in range(0,4):
            down = down * grid[i+k][j]
            accross = accross * grid[i][j+k]
            diag_dr = diag_dr * grid[i+k][j+k]
            diag_ur = diag_ur * grid[i+3-k][j+k]
        results.update({'R{:02}C{:02}-{}'.format(i+1,j+1,'D'):down})
        results.update({'R{:02}C{:02}-{}'.format(i+1,j+1,'A'):accross})
        results.update({'R{:02}C{:02}-{}'.format(i+1,j+1,'Dr'):diag_dr})
        results.update({'R{:02}C{:02}-{}'.format(i+1+3,j+1,'Ur'):diag_ur})

print('The highest product is {} found at location {}.'.format(max(results.values()),[k for k,v in results.items() if v == max(results.values())][0]))

The highest product is 70600674 found at location R16C04-Ur.


# Highly divisible triangular number
## Problem 12 
The sequence of triangle numbers is generated by adding the natural numbers. So the 7th triangle number would be 1 + 2 + 3 + 4 + 5 + 6 + 7 = 28. The first ten terms would be:

1, 3, 6, 10, 15, 21, 28, 36, 45, 55, ...

Let us list the factors of the first seven triangle numbers:

 1: 1
 3: 1,3
 6: 1,2,3,6
10: 1,2,5,10
15: 1,3,5,15
21: 1,3,7,21
28: 1,2,4,7,14,28
We can see that 28 is the first triangle number to have over five divisors.

What is the value of the first triangle number to have over five hundred divisors?

In [156]:
def ordinal(num):
    suffixes = {1: 'st', 2: 'nd', 3: 'rd'}
    # I'm checking for 10-20 because those are the digits that
    # don't follow the normal counting scheme. 
    if 10 <= num % 100 <= 20:
        suffix = 'th'
    else:
        # the second parameter is a default.
        suffix = suffixes.get(num % 10, 'th')
    return str(num) + suffix


In [189]:
n = 1
triangles = {}
divisors = []

while len(divisors) <= 500:
    num = 0
    divisors = []
    
    for i in range(1,n+1):
        num += i
    
    i = 1
    while (len(divisors) == 0) or (i < divisors[-1]) :
        if num%i == 0:
            divisors.append(i)
            divisors.append(int(num/i))
        i += 1
        
    triangles.update({n:(num,len(divisors), divisors.sort())})   
    
    if len(divisors) > 500:
        break
    
    n += 1

print('The {} triangular number is {}. It has {}  divisors : {}'.format(ordinal(n),num, len(divisors), divisors))

The 12375th triangular number is 76576500. It has 576  divisors : [1, 2, 3, 4, 5, 6, 7, 9, 10, 11, 12, 13, 14, 15, 17, 18, 20, 21, 22, 25, 26, 28, 30, 33, 34, 35, 36, 39, 42, 44, 45, 50, 51, 52, 55, 60, 63, 65, 66, 68, 70, 75, 77, 78, 84, 85, 90, 91, 99, 100, 102, 105, 110, 117, 119, 125, 126, 130, 132, 140, 143, 150, 153, 154, 156, 165, 170, 175, 180, 182, 187, 195, 198, 204, 210, 220, 221, 225, 231, 234, 238, 250, 252, 255, 260, 273, 275, 286, 300, 306, 308, 315, 325, 330, 340, 350, 357, 364, 374, 375, 385, 390, 396, 420, 425, 429, 442, 450, 455, 462, 468, 476, 495, 500, 510, 525, 546, 550, 561, 572, 585, 595, 612, 630, 650, 660, 663, 693, 700, 714, 715, 748, 750, 765, 770, 780, 819, 825, 850, 858, 875, 884, 900, 910, 924, 935, 975, 990, 1001, 1020, 1050, 1071, 1092, 1100, 1105, 1122, 1125, 1155, 1170, 1190, 1260, 1275, 1287, 1300, 1309, 1326, 1365, 1375, 1386, 1428, 1430, 1500, 1530, 1540, 1547, 1575, 1625, 1638, 1650, 1683, 1700, 1716, 1750, 1785, 1820, 1870, 1925, 1950, 1980, 1989

# Large sum
## Problem 13 
Work out the first ten digits of the sum of the following one-hundred 50-digit numbers.

37107287533902102798797998220837590246510135740250
46376937677490009712648124896970078050417018260538
74324986199524741059474233309513058123726617309629
91942213363574161572522430563301811072406154908250
23067588207539346171171980310421047513778063246676
89261670696623633820136378418383684178734361726757
28112879812849979408065481931592621691275889832738
44274228917432520321923589422876796487670272189318
47451445736001306439091167216856844588711603153276
70386486105843025439939619828917593665686757934951
62176457141856560629502157223196586755079324193331
64906352462741904929101432445813822663347944758178
92575867718337217661963751590579239728245598838407
58203565325359399008402633568948830189458628227828
80181199384826282014278194139940567587151170094390
35398664372827112653829987240784473053190104293586
86515506006295864861532075273371959191420517255829
71693888707715466499115593487603532921714970056938
54370070576826684624621495650076471787294438377604
53282654108756828443191190634694037855217779295145
36123272525000296071075082563815656710885258350721
45876576172410976447339110607218265236877223636045
17423706905851860660448207621209813287860733969412
81142660418086830619328460811191061556940512689692
51934325451728388641918047049293215058642563049483
62467221648435076201727918039944693004732956340691
15732444386908125794514089057706229429197107928209
55037687525678773091862540744969844508330393682126
18336384825330154686196124348767681297534375946515
80386287592878490201521685554828717201219257766954
78182833757993103614740356856449095527097864797581
16726320100436897842553539920931837441497806860984
48403098129077791799088218795327364475675590848030
87086987551392711854517078544161852424320693150332
59959406895756536782107074926966537676326235447210
69793950679652694742597709739166693763042633987085
41052684708299085211399427365734116182760315001271
65378607361501080857009149939512557028198746004375
35829035317434717326932123578154982629742552737307
94953759765105305946966067683156574377167401875275
88902802571733229619176668713819931811048770190271
25267680276078003013678680992525463401061632866526
36270218540497705585629946580636237993140746255962
24074486908231174977792365466257246923322810917141
91430288197103288597806669760892938638285025333403
34413065578016127815921815005561868836468420090470
23053081172816430487623791969842487255036638784583
11487696932154902810424020138335124462181441773470
63783299490636259666498587618221225225512486764533
67720186971698544312419572409913959008952310058822
95548255300263520781532296796249481641953868218774
76085327132285723110424803456124867697064507995236
37774242535411291684276865538926205024910326572967
23701913275725675285653248258265463092207058596522
29798860272258331913126375147341994889534765745501
18495701454879288984856827726077713721403798879715
38298203783031473527721580348144513491373226651381
34829543829199918180278916522431027392251122869539
40957953066405232632538044100059654939159879593635
29746152185502371307642255121183693803580388584903
41698116222072977186158236678424689157993532961922
62467957194401269043877107275048102390895523597457
23189706772547915061505504953922979530901129967519
86188088225875314529584099251203829009407770775672
11306739708304724483816533873502340845647058077308
82959174767140363198008187129011875491310547126581
97623331044818386269515456334926366572897563400500
42846280183517070527831839425882145521227251250327
55121603546981200581762165212827652751691296897789
32238195734329339946437501907836945765883352399886
75506164965184775180738168837861091527357929701337
62177842752192623401942399639168044983993173312731
32924185707147349566916674687634660915035914677504
99518671430235219628894890102423325116913619626622
73267460800591547471830798392868535206946944540724
76841822524674417161514036427982273348055556214818
97142617910342598647204516893989422179826088076852
87783646182799346313767754307809363333018982642090
10848802521674670883215120185883543223812876952786
71329612474782464538636993009049310363619763878039
62184073572399794223406235393808339651327408011116
66627891981488087797941876876144230030984490851411
60661826293682836764744779239180335110989069790714
85786944089552990653640447425576083659976645795096
66024396409905389607120198219976047599490197230297
64913982680032973156037120041377903785566085089252
16730939319872750275468906903707539413042652315011
94809377245048795150954100921645863754710598436791
78639167021187492431995700641917969777599028300699
15368713711936614952811305876380278410754449733078
40789923115535562561142322423255033685442488917353
44889911501440648020369068063960672322193204149535
41503128880339536053299340368006977710650566631954
81234880673210146739058568557934581403627822703280
82616570773948327592232845941706525094512325230608
22918802058777319719839450180888072429661980811197
77158542502016545090413245809786882778948721859617
72107838435069186155435662884062257473692284509516
20849603980134001723930671666823555245252804609722
53503534226472524250874054075591789781264330331690

In [304]:
numbers = []
numbers.append('37107287533902102798797998220837590246510135740250')
numbers.append('46376937677490009712648124896970078050417018260538')
numbers.append('74324986199524741059474233309513058123726617309629')
numbers.append('91942213363574161572522430563301811072406154908250')
numbers.append('23067588207539346171171980310421047513778063246676')
numbers.append('89261670696623633820136378418383684178734361726757')
numbers.append('28112879812849979408065481931592621691275889832738')
numbers.append('44274228917432520321923589422876796487670272189318')
numbers.append('47451445736001306439091167216856844588711603153276')
numbers.append('70386486105843025439939619828917593665686757934951')
numbers.append('62176457141856560629502157223196586755079324193331')
numbers.append('64906352462741904929101432445813822663347944758178')
numbers.append('92575867718337217661963751590579239728245598838407')
numbers.append('58203565325359399008402633568948830189458628227828')
numbers.append('80181199384826282014278194139940567587151170094390')
numbers.append('35398664372827112653829987240784473053190104293586')
numbers.append('86515506006295864861532075273371959191420517255829')
numbers.append('71693888707715466499115593487603532921714970056938')
numbers.append('54370070576826684624621495650076471787294438377604')
numbers.append('53282654108756828443191190634694037855217779295145')
numbers.append('36123272525000296071075082563815656710885258350721')
numbers.append('45876576172410976447339110607218265236877223636045')
numbers.append('17423706905851860660448207621209813287860733969412')
numbers.append('81142660418086830619328460811191061556940512689692')
numbers.append('51934325451728388641918047049293215058642563049483')
numbers.append('62467221648435076201727918039944693004732956340691')
numbers.append('15732444386908125794514089057706229429197107928209')
numbers.append('55037687525678773091862540744969844508330393682126')
numbers.append('18336384825330154686196124348767681297534375946515')
numbers.append('80386287592878490201521685554828717201219257766954')
numbers.append('78182833757993103614740356856449095527097864797581')
numbers.append('16726320100436897842553539920931837441497806860984')
numbers.append('48403098129077791799088218795327364475675590848030')
numbers.append('87086987551392711854517078544161852424320693150332')
numbers.append('59959406895756536782107074926966537676326235447210')
numbers.append('69793950679652694742597709739166693763042633987085')
numbers.append('41052684708299085211399427365734116182760315001271')
numbers.append('65378607361501080857009149939512557028198746004375')
numbers.append('35829035317434717326932123578154982629742552737307')
numbers.append('94953759765105305946966067683156574377167401875275')
numbers.append('88902802571733229619176668713819931811048770190271')
numbers.append('25267680276078003013678680992525463401061632866526')
numbers.append('36270218540497705585629946580636237993140746255962')
numbers.append('24074486908231174977792365466257246923322810917141')
numbers.append('91430288197103288597806669760892938638285025333403')
numbers.append('34413065578016127815921815005561868836468420090470')
numbers.append('23053081172816430487623791969842487255036638784583')
numbers.append('11487696932154902810424020138335124462181441773470')
numbers.append('63783299490636259666498587618221225225512486764533')
numbers.append('67720186971698544312419572409913959008952310058822')
numbers.append('95548255300263520781532296796249481641953868218774')
numbers.append('76085327132285723110424803456124867697064507995236')
numbers.append('37774242535411291684276865538926205024910326572967')
numbers.append('23701913275725675285653248258265463092207058596522')
numbers.append('29798860272258331913126375147341994889534765745501')
numbers.append('18495701454879288984856827726077713721403798879715')
numbers.append('38298203783031473527721580348144513491373226651381')
numbers.append('34829543829199918180278916522431027392251122869539')
numbers.append('40957953066405232632538044100059654939159879593635')
numbers.append('29746152185502371307642255121183693803580388584903')
numbers.append('41698116222072977186158236678424689157993532961922')
numbers.append('62467957194401269043877107275048102390895523597457')
numbers.append('23189706772547915061505504953922979530901129967519')
numbers.append('86188088225875314529584099251203829009407770775672')
numbers.append('11306739708304724483816533873502340845647058077308')
numbers.append('82959174767140363198008187129011875491310547126581')
numbers.append('97623331044818386269515456334926366572897563400500')
numbers.append('42846280183517070527831839425882145521227251250327')
numbers.append('55121603546981200581762165212827652751691296897789')
numbers.append('32238195734329339946437501907836945765883352399886')
numbers.append('75506164965184775180738168837861091527357929701337')
numbers.append('62177842752192623401942399639168044983993173312731')
numbers.append('32924185707147349566916674687634660915035914677504')
numbers.append('99518671430235219628894890102423325116913619626622')
numbers.append('73267460800591547471830798392868535206946944540724')
numbers.append('76841822524674417161514036427982273348055556214818')
numbers.append('97142617910342598647204516893989422179826088076852')
numbers.append('87783646182799346313767754307809363333018982642090')
numbers.append('10848802521674670883215120185883543223812876952786')
numbers.append('71329612474782464538636993009049310363619763878039')
numbers.append('62184073572399794223406235393808339651327408011116')
numbers.append('66627891981488087797941876876144230030984490851411')
numbers.append('60661826293682836764744779239180335110989069790714')
numbers.append('85786944089552990653640447425576083659976645795096')
numbers.append('66024396409905389607120198219976047599490197230297')
numbers.append('64913982680032973156037120041377903785566085089252')
numbers.append('16730939319872750275468906903707539413042652315011')
numbers.append('94809377245048795150954100921645863754710598436791')
numbers.append('78639167021187492431995700641917969777599028300699')
numbers.append('15368713711936614952811305876380278410754449733078')
numbers.append('40789923115535562561142322423255033685442488917353')
numbers.append('44889911501440648020369068063960672322193204149535')
numbers.append('41503128880339536053299340368006977710650566631954')
numbers.append('81234880673210146739058568557934581403627822703280')
numbers.append('82616570773948327592232845941706525094512325230608')
numbers.append('22918802058777319719839450180888072429661980811197')
numbers.append('77158542502016545090413245809786882778948721859617')
numbers.append('72107838435069186155435662884062257473692284509516')
numbers.append('20849603980134001723930671666823555245252804609722')
numbers.append('53503534226472524250874054075591789781264330331690')

In [326]:
def arraysum(n1,n2):
    if len(n1) < len(n2):
        n1 = [0]*(len(n2) - len(n1))+n1
    elif len(n2) < len(n1):
        n2 = [0]*(len(n1) - len(n2))+n2

    buffer = n1
    for i in range(1,len(buffer)+1):

        buffer[-i] = buffer[-i] + n2[-i]
        if buffer[-i] > 9 and i == len(buffer):
            buffer = [0] + buffer
        if buffer[-i] > 9:     
            buffer[-i-1] = buffer[-i-1] + int((buffer[-i]-buffer[-i]%10)/10)  
            buffer[-i] = buffer[-i]%10

    while buffer[0] == 0:
        buffer.pop(0)
   
    print(''.join(map(str,buffer)))
    return buffer

In [332]:
result = [0]
print(''.join(map(str,result)))

for numstr in numbers:
    result = arraysum(result, [int(num) for num in numstr])
    
print(''.join(map(str,result)))
print()
print(''.join(map(str,result[0:10])))

0
5537376230390876637302048746832985971773659831892672

5537376230


# Longest Collatz sequence
## Problem 14 
The following iterative sequence is defined for the set of positive integers:

n → n/2 (n is even)
n → 3n + 1 (n is odd)

Using the rule above and starting with 13, we generate the following sequence:

13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1
It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

Which starting number, under one million, produces the longest chain?

NOTE: Once the chain starts the terms are allowed to go above one million.

In [339]:
def collatz(num):
    seq = [num]
    while seq[-1] > 1:
        if {collatz_seq}
        if seq[-1]%2 == 0:
            seq.append(int(seq[-1]/2))
        else:
            seq.append(seq[-1]*3 + 1)
    return seq
    

In [385]:
collatz(9)

[9, 28, 14, 7, 22, 11, 34, 17, 52, 26, 13, 40, 20, 10, 5, 16, 8, 4, 2, 1]

In [424]:
collatz_seq = {1:[1]}
longest_seq = 1

for num in range(2,10**6):
    seq = [num]
    while seq[-1] > 1:
        if seq[-1] in collatz_seq.keys():
            seq = seq +  collatz_seq[seq[-1]][1:]        
        elif seq[-1]%2 == 0:
            seq.append(int(seq[-1]/2))
        else:
            seq.append(seq[-1]*3 + 1)
    
    if len(seq) > len(collatz_seq[longest_seq]):
        longest_seq = num
        
    #print('{:2} {}'.format(str(num),','.join(map(str,seq)))) 
    
    #old code. amended to imporove calculation speed by saving intermediate points
    #collatz_seq.update({num:seq})
    
    while seq[0] not in collatz_seq.keys():
        collatz_seq.update({seq[0]:seq})
        seq = seq[1:]
        
    
    
    
print('The longest sequence is {} with length {}'.format(longest_seq,len(collatz_seq[longest_seq])))        
print(collatz_seq[longest_seq])
            


The longest sequence is 837799 with length 525
[837799, 2513398, 1256699, 3770098, 1885049, 5655148, 2827574, 1413787, 4241362, 2120681, 6362044, 3181022, 1590511, 4771534, 2385767, 7157302, 3578651, 10735954, 5367977, 16103932, 8051966, 4025983, 12077950, 6038975, 18116926, 9058463, 27175390, 13587695, 40763086, 20381543, 61144630, 30572315, 91716946, 45858473, 137575420, 68787710, 34393855, 103181566, 51590783, 154772350, 77386175, 232158526, 116079263, 348237790, 174118895, 522356686, 261178343, 783535030, 391767515, 1175302546, 587651273, 1762953820, 881476910, 440738455, 1322215366, 661107683, 1983323050, 991661525, 2974984576, 1487492288, 743746144, 371873072, 185936536, 92968268, 46484134, 23242067, 69726202, 34863101, 104589304, 52294652, 26147326, 13073663, 39220990, 19610495, 58831486, 29415743, 88247230, 44123615, 132370846, 66185423, 198556270, 99278135, 297834406, 148917203, 446751610, 223375805, 670127416, 335063708, 167531854, 83765927, 251297782, 125648891, 376946674, 1

# Lattice paths
## Problem 15 
Starting in the top left corner of a 2×2 grid, and only being able to move to the right and down, there are exactly 6 routes to the bottom right corner.

<img src=https://projecteuler.net/project/images/p015.gif>

How many such routes are there through a 20×20 grid?

In [487]:
matrix_size = 21

paths = [[0 for x in range(matrix_size)] for x in range(matrix_size)]

for x in range(matrix_size-1,-1,-1):
    for y in range(matrix_size-1,-1,-1): 
        if x == matrix_size-1:
            paths[x][y] = 1
        elif y == matrix_size-1:
            paths[x][y] = 1
        else:
            paths[x][y] = paths[x+1][y] + paths[x][y+1]
                    
print('There are {} paths from top left to bottom right'.format(paths[0][0]))
print()
#for x in paths:
#    for y in x:
#        print('{:3} '.format(y), end='')
#    print()

There are 137846528820 paths from top left to bottom right



In [496]:
#Simpler result for those with better math 

def countRoutes(n):
    result = 1
    for i in range(1,n+1):
        result = result * (n + i)/i
    return result

countRoutes(2)

6.0

# Power digit sum
## Problem 16 
215 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.

What is the sum of the digits of the number 21000?

In [553]:
def arraysum(n1,n2):
    if len(n1) < len(n2):
        n1 = [0]*(len(n2) - len(n1))+n1
    elif len(n2) < len(n1):
        n2 = [0]*(len(n1) - len(n2))+n2

    buffer = n1
    for i in range(1,len(buffer)+1):

        buffer[-i] = buffer[-i] + n2[-i]
        if buffer[-i] > 9 and i == len(buffer):
            buffer = [0] + buffer
        if buffer[-i] > 9:     
            buffer[-i-1] = buffer[-i-1] + int((buffer[-i]-buffer[-i]%10)/10)  
            buffer[-i] = buffer[-i]%10
            
    while buffer[0] == 0:
        buffer.pop(0)
   
    #print(''.join(map(str,buffer)))
    return buffer

In [554]:
power = 1000

result = [1]
for i in range(power):
    result = arraysum(list(result), list(result)) # have to use list(result) to avoid result being passed as reference 
    
print('2 to the power of {} is :\n\n{}\n'.format(power,''.join(map(str,result))))
print('This number is {} digits long\n'.format(len(result)))
print('The sum of these digits is {}'.format(sum(result)))



2 to the power of 1000 is :

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376

This number is 302 digits long

The sum of these digits is 1366


# Number letter counts
## Problem 17 
If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?


NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with British usage.

In [602]:
num2words1 = {0: 'Zero', 1: 'One', 2: 'Two', 3: 'Three', 4: 'Four', 5: 'Five', \
            6: 'Six', 7: 'Seven', 8: 'Eight', 9: 'Nine', 10: 'Ten', \
            11: 'Eleven', 12: 'Twelve', 13: 'Thirteen', 14: 'Fourteen', \
            15: 'Fifteen', 16: 'Sixteen', 17: 'Seventeen', 18: 'Eighteen', 19: 'Nineteen'}
num2words2 = {20:'Twenty', 30:'Thirty', 40:'Forty', 50:'Fifty', 60:'Sixty', 70:'Seventy', 80:'Eighty', 90:'Ninety'}

In [646]:
def numtowords(number):
    if number < 20:
        return num2words1[number]
    elif number < 100 and number%10 == 0:
        return num2words2[number]
    elif number < 100:
        return num2words2[number-number%10] + ' ' + num2words1[number%10]
    elif number < 1000 and number%100 == 0:
        return num2words1[number//100] + ' Hundred'
    elif number < 1000:
        return num2words1[number//100] + ' Hundred and ' + numtowords(number%100)
    elif number < 1000000:    
        if number%1000 == 0:
            return numtowords(number//1000) + ' Thousand'
        if number%1000 < 100:
            return numtowords(number//1000) + ' Thousand and ' + numtowords(number%1000)
        else:
            return numtowords(number//1000) + ' Thousand, ' + numtowords(number%1000)
    else:      
        return number


In [655]:
length = 0
for i in range(1,1001):
    length += len([letter for letter in numtowords(i) if letter not in [' ', ',']]) 
           
print(length)

21124


# Maximum path sum I
## Problem 18

By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.

3
7 4
2 4 6
8 5 9 3

That is, 3 + 7 + 4 + 9 = 23.

Find the maximum total from top to bottom of the triangle below:

75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23

NOTE: As there are only 16384 routes, it is possible to solve this problem by trying every route. However, Problem 67, is the same challenge with a triangle containing one-hundred rows; it cannot be solved by brute force, and requires a clever method! ;o)

In [714]:
triangle = []

triangle.append(list(map(int,'75'.split())))
triangle.append(list(map(int,'95 64'.split())))
triangle.append(list(map(int,'17 47 82'.split())))
triangle.append(list(map(int,'18 35 87 10'.split())))
triangle.append(list(map(int,'20 04 82 47 65'.split())))
triangle.append(list(map(int,'19 01 23 75 03 34'.split())))
triangle.append(list(map(int,'88 02 77 73 07 63 67'.split())))
triangle.append(list(map(int,'99 65 04 28 06 16 70 92'.split())))
triangle.append(list(map(int,'41 41 26 56 83 40 80 70 33'.split())))
triangle.append(list(map(int,'41 48 72 33 47 32 37 16 94 29'.split())))
triangle.append(list(map(int,'53 71 44 65 25 43 91 52 97 51 14'.split())))
triangle.append(list(map(int,'70 11 33 28 77 73 17 78 39 68 17 57'.split())))
triangle.append(list(map(int,'91 71 52 38 17 14 91 43 58 50 27 29 48'.split())))
triangle.append(list(map(int,'63 66 04 68 89 53 67 30 73 16 69 87 40 31'.split())))
triangle.append(list(map(int,'04 62 98 27 23 09 70 98 73 93 38 53 60 04 23'.split())))

# (value, maxsum, right)

for i in range(len(triangle)-1,-1,-1):
    for j in range(len(triangle[i])):
        if i == len(triangle)-1:
            triangle[i][j] = (triangle[i][j],triangle[i][j],)
        else:
            if triangle[i+1][j][1] > triangle[i+1][j+1][1]:
                triangle[i][j] = (triangle[i][j],triangle[i][j]+triangle[i+1][j][1],False)
            else:
                triangle[i][j] = (triangle[i][j],triangle[i][j]+triangle[i+1][j+1][1],True)                   
        
print('The highest path with a value of {} is :'.format(triangle[0][0][1]))
print()

r = 0
bestpath = []
for i in range(len(triangle)-1):
    print(triangle[i][r][0])
    if triangle[i][r][2]:
        r += 1
        

The highest path with a value of 1074 is :

75
64
82
87
82
75
73
28
83
32
91
78
58
73


# Counting Sundays
# Problem 19 
You are given the following information, but you may prefer to do some research for yourself.

1 Jan 1900 was a Monday.
Thirty days has September,
April, June and November.
All the rest have thirty-one,
Saving February alone,
Which has twenty-eight, rain or shine.
And on leap years, twenty-nine.
A leap year occurs on any year evenly divisible by 4, but not on a century unless it is divisible by 400.
How many Sundays fell on the first of the month during the twentieth century (1 Jan 1901 to 31 Dec 2000)?

In [68]:
daymap = {0:'MON',1:'TUE',2:'WED',3:'THU',4:'FRI',5:'SAT',6:'SUN'}
month = {0:'',1:'',2:'',3:'',4:'',5:'',6:'',7:'',8:'',9:'',10:'',10:''}

def weekday(date):    
    return daymap[date%7]

def yearstart(year):
    return (year - 1901) * 365 + len([i for i in range(1901,year) if i%4 == 0 and i%1000 != 0])

def monthstart(year, month):
    while month > 12:
        year += 1
        month -= 12
    value = yearstart(year) 
    for i in range (1,month):
        if i in [4,6,9,11]:
            value += 30
        elif i == 2:
            if year%4 == 0 and year/1000 != 0:
                value += 29
            else:
                value += 28
        else:
            value += 31

    return value
 
results = []    
for i in range(1901,2000):
    for j in range(1,13):
        if weekday(monthstart(i,j)) == 'SUN':
            results.append((i,j,1,weekday(monthstart(i,j)),monthstart(i,j+1)-monthstart(i,j)))
            
print(len(results))

171


In [70]:
results

[(1901, 4, 1, 'SUN', 30),
 (1901, 7, 1, 'SUN', 31),
 (1902, 9, 1, 'SUN', 30),
 (1902, 12, 1, 'SUN', 31),
 (1903, 6, 1, 'SUN', 30),
 (1904, 2, 1, 'SUN', 29),
 (1904, 8, 1, 'SUN', 31),
 (1905, 5, 1, 'SUN', 31),
 (1906, 1, 1, 'SUN', 31),
 (1906, 10, 1, 'SUN', 31),
 (1907, 4, 1, 'SUN', 30),
 (1907, 7, 1, 'SUN', 31),
 (1908, 6, 1, 'SUN', 30),
 (1909, 2, 1, 'SUN', 28),
 (1909, 3, 1, 'SUN', 31),
 (1909, 11, 1, 'SUN', 30),
 (1910, 8, 1, 'SUN', 31),
 (1911, 5, 1, 'SUN', 31),
 (1912, 1, 1, 'SUN', 31),
 (1912, 4, 1, 'SUN', 30),
 (1912, 7, 1, 'SUN', 31),
 (1913, 9, 1, 'SUN', 30),
 (1913, 12, 1, 'SUN', 31),
 (1914, 6, 1, 'SUN', 30),
 (1915, 2, 1, 'SUN', 28),
 (1915, 3, 1, 'SUN', 31),
 (1915, 11, 1, 'SUN', 30),
 (1916, 5, 1, 'SUN', 31),
 (1917, 1, 1, 'SUN', 31),
 (1917, 10, 1, 'SUN', 31),
 (1918, 4, 1, 'SUN', 30),
 (1918, 7, 1, 'SUN', 31),
 (1919, 9, 1, 'SUN', 30),
 (1919, 12, 1, 'SUN', 31),
 (1920, 3, 1, 'SUN', 31),
 (1920, 11, 1, 'SUN', 30),
 (1921, 8, 1, 'SUN', 31),
 (1922, 5, 1, 'SUN', 31),
 (19

# Factorial digit sum
## Problem 20 
n! means n × (n − 1) × ... × 3 × 2 × 1

For example, 10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800,
and the sum of the digits in the number 10! is 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27.

Find the sum of the digits in the number 100!

In [85]:
def factorial(number):
    result = 1
    for i in range(number,0,-1):
        result = result * i
    return result 

number = 100
print('!{} is {}'.format(number,factorial(number)))
print('The sum of digits is {}'.format(sum(map(int,[letter for letter in str(factorial(number))]))))

!100 is 93326215443944152681699238856266700490715968264381621468592963895217599993229915608941463976156518286253697920827223758251185210916864000000000000000000000000
The sum of digits is 648


# Amicable numbers
## Problem 21 
Let d(n) be defined as the sum of proper divisors of n (numbers less than n which divide evenly into n).
If d(a) = b and d(b) = a, where a ≠ b, then a and b are an amicable pair and each of a and b are called amicable numbers.

For example, the proper divisors of 220 are 1, 2, 4, 5, 10, 11, 20, 22, 44, 55 and 110; therefore d(220) = 284. The proper divisors of 284 are 1, 2, 4, 71 and 142; so d(284) = 220.

Evaluate the sum of all the amicable numbers under 10000.

In [221]:
divisors = {}
d = {}

for i in range(2,10000):
    
    
    divisors.update({i:[1]})
    
    j = 2
    while j < i ** 0.5:
        if i%j == 0:
            divisors[i].append(j)
            if j**2 != i and j < i:
                divisors[i].append(i//j)
        j += 1
    

    #divisors[i].sort()
    d.update({i:sum(divisors[i])})
    

    
amicable = [k2 for k1, v1 in d.items() for k2, v2 in d.items() if k1==v2 and k2==v1 and k1!=v1]
print('There are {} amicable numbers below {}. These are :\n\n{}\n'.format(len(amicable),max(d.keys())+1,amicable))
print('The sum of these is {}.'.format(sum(amicable)))

There are 10 amicable numbers below 10000. These are :

[284, 220, 1210, 1184, 2924, 2620, 5564, 5020, 6368, 6232]

The sum of these is 31626.


# Names scores
## Problem 22 
Using names.txt (right click and 'Save Link/Target As...'), a 46K text file containing over five-thousand first names, begin by sorting it into alphabetical order. Then working out the alphabetical value for each name, multiply this value by its alphabetical position in the list to obtain a name score.

For example, when the list is sorted into alphabetical order, COLIN, which is worth 3 + 15 + 12 + 9 + 14 = 53, is the 938th name in the list. So, COLIN would obtain a score of 938 × 53 = 49714.

What is the total of all the name scores in the file?

In [250]:
inputfile = open('C:\\Users\\alexp\\Downloads\\Data Science\\Project Euler\\p022_names.txt','r')
names = [name[1:-1] for name in inputfile.read().split(',')]
inputfile.close()
names.sort()

names = {name: sum([ord(letter)-64 for letter in name])*(names.index(name)+1) for name in names}

print('There are {} names with a total value of {}.'.format(len(names), sum(names.values())))


There are 5163 names with a total value of 871198282.


# Non-abundant sums
## Problem 23 
A perfect number is a number for which the sum of its proper divisors is exactly equal to the number. For example, the sum of the proper divisors of 28 would be 1 + 2 + 4 + 7 + 14 = 28, which means that 28 is a perfect number.

A number n is called deficient if the sum of its proper divisors is less than n and it is called abundant if this sum exceeds n.

As 12 is the smallest abundant number, 1 + 2 + 3 + 4 + 6 = 16, the smallest number that can be written as the sum of two abundant numbers is 24. By mathematical analysis, it can be shown that all integers greater than 28123 can be written as the sum of two abundant numbers. However, this upper limit cannot be reduced any further by analysis even though it is known that the greatest number that cannot be expressed as the sum of two abundant numbers is less than this limit.

Find the sum of all the positive integers which cannot be written as the sum of two abundant numbers.

In [90]:
divisors = {}
abundant = []

maxcalc = 28123

for i in range(2,maxcalc):
    divisors.update({i:[1]})
   
    j = 2
    while j <= i ** 0.5:
        if i%j == 0:
            divisors[i].append(j)
            if j**2 != i and j < i:
                divisors[i].append(i//j)
        j += 1
    divisors[i].sort()
    
    if sum(divisors[i]) > i:
        abundant.append(i)

        abundantset = set(abundant)

def is_abundant_sum(n):
    for i in abundant:
        if i > n:  # assume "abundants" is ordered
            return False
        if (n - i) in abundantset:
            return True
    return False

integer2a  = [x for x in range(1, maxcalc + 1) if is_abundant_sum(x)]    
integernot2a = [x for x in range(1, maxcalc + 1) if x not in integer2a] 

print('There are {} integers that cannot be expressed as the sum of two abundant numbers'.format(len(integernot2a)))
print('The sum of these is {}.'.format(sum(integernot2a)))

There are 1456 integers that cannot be expressed as the sum of two abundant numbers
The sum of these is 4179871.


# Lexicographic permutations
## Problem 24 
A permutation is an ordered arrangement of objects. For example, 3124 is one possible permutation of the digits 1, 2, 3 and 4. If all of the permutations are listed numerically or alphabetically, we call it lexicographic order. The lexicographic permutations of 0, 1 and 2 are:

012   021   102   120   201   210

What is the millionth lexicographic permutation of the digits 0, 1, 2, 3, 4, 5, 6, 7, 8 and 9?

In [93]:
#options = [n for n in range(0,10)]


In [18]:
def recurse_permutations(fixeditems, floatingitems):
    
    for i in floatingitems:
        if len(floatingitems) == 1:
            permutations.append(fixeditems+[i])
        else:
            newfix = list(fixeditems)
            newfix.append(i)
            newfloat = list(floatingitems)
            newfloat.pop(newfloat.index(i))
        
            recurse_permutations(list(newfix),list(newfloat))
            if len(permutations) >= 1000000:
                break
        
permutations = []
recurse_permutations(list([]), list(range(0,10)))

print(len(permutations))
print(''.join(map(str,permutations[-1])))

1000000
2783915460


# 1000-digit Fibonacci number
## Problem 25 
The Fibonacci sequence is defined by the recurrence relation:

Fn = Fn−1 + Fn−2, where F1 = 1 and F2 = 1.
Hence the first 12 terms will be:

F1 = 1
F2 = 1
F3 = 2
F4 = 3
F5 = 5
F6 = 8
F7 = 13
F8 = 21
F9 = 34
F10 = 55
F11 = 89
F12 = 144
The 12th term, F12, is the first term to contain three digits.

What is the index of the first term in the Fibonacci sequence to contain 1000 digits?

In [253]:
def arraysum(n1,n2):
    if len(n1) < len(n2):
        n1 = [0]*(len(n2) - len(n1))+n1
    elif len(n2) < len(n1):
        n2 = [0]*(len(n1) - len(n2))+n2

    buffer = n1
    for i in range(1,len(buffer)+1):

        buffer[-i] = buffer[-i] + n2[-i]
        if buffer[-i] > 9 and i == len(buffer):
            buffer = [0] + buffer
        if buffer[-i] > 9:     
            buffer[-i-1] = buffer[-i-1] + int((buffer[-i]-buffer[-i]%10)/10)  
            buffer[-i] = buffer[-i]%10
            
    while buffer[0] == 0:
        buffer.pop(0)
   
    #print(''.join(map(str,buffer)))
    return buffer

In [268]:
#{N, fib[n], fib[n-1]}

target = 1000

fibcalc = (2,[1],[1])
print(fibcalc)

while len(fibcalc[1])<target:
    fibcalc = (fibcalc[0]+1,arraysum(list(fibcalc[1]), list(fibcalc[2])), list(fibcalc[1]))
    #print(fibcalc)     

print('F{} is the first Fibonacci number to contain {} digits. It is :\n\n{}\n'.format(fibcalc[0],target,''.join(map(str,fibcalc[1]))))

(2, [1], [1])
F4782 is the first Fibonacci number to contain 1000 digits. It is :

10700662663827589367649805844573968850836838966321516650132352033753145206046940406218891475824897926578046948881775919574843364666725699595129960304612627480924821861440694330512347744427502737817530875793916661921492591867595539664228371489431130746995034395470019854326097230672901928705264472437261177158218255484911205250132014786129659313817922355596574520395061375514678375432291196021299340482607061753977068470682028954869026661854351245219003694806413574474709117076197669456910700980243934396174741037369125032313655321647736970231677550515951735184605799549194109677783732296657965816465139034881542563101842241902598460880001101862555502454939371136516570394476295847145485234259504285824253060835444354282126110089928637950480068943303097732178348645431132057656598684562886168087186938352973506439862976406600007235629179052070511640776148124918858309459405666883391093509444565763576661516193177537928

# Reciprocal cycles
## Problem 26 
A unit fraction contains 1 in the numerator. The decimal representation of the unit fractions with denominators 2 to 10 are given:

1/2	= 	0.5
1/3	= 	0.(3)
1/4	= 	0.25
1/5	= 	0.2
1/6	= 	0.1(6)
1/7	= 	0.(142857)
1/8	= 	0.125
1/9	= 	0.(1)
1/10	= 	0.1
Where 0.1(6) means 0.166666..., and has a 1-digit recurring cycle. It can be seen that 1/7 has a 6-digit recurring cycle.

Find the value of d < 1000 for which 1/d contains the longest recurring cycle in its decimal fraction part.

In [279]:
def reciprocal_cycle(d):
    output = [1//d]
    remainder = [1%7]

    while remainder[-1] != 0:
        output.append(remainder[-1]*10//d)

        if output[-1] == 0 and remainder[-1]*10%d == 0:
            return (d, 0, output, [])
        elif remainder[-1]*10%d in remainder:
            length = len(remainder) - remainder.index(remainder[-1]*10%d)
            return (d, length, output[0:-length], output[-length:])
        else:
            remainder.append(remainder[-1]*10%d)
    return output  

In [280]:
reciprocal_cycles = []

longestcycle = 0
longest_at_d = 1

for i in range(2,1000):
    reciprocal_cycles.append(reciprocal_cycle(i))
    if reciprocal_cycles[-1][1] > longestcycle:
        longestcycle = reciprocal_cycles[-1][1]
        longest_at_d = i
        
print(longestcycle)
print(longest_at_d)

982
983


# Quadratic primes
## Problem 27 
Euler discovered the remarkable quadratic formula:

n2+n+41
It turns out that the formula will produce 40 primes for the consecutive integer values 0≤n≤39. However, when n=40,402+40+41=40(40+1)+41 is divisible by 41, and certainly when n=41,412+41+41 is clearly divisible by 41.

The incredible formula n2−79n+1601 was discovered, which produces 80 primes for the consecutive values 0≤n≤79. The product of the coefficients, −79 and 1601, is −126479.

Considering quadratics of the form:

n2+an+b, where |a|<1000 and |b|≤1000

where |n| is the modulus/absolute value of n
e.g. |11|=11 and |−4|=4
Find the product of the coefficients, a and b, for the quadratic expression that produces the maximum number of primes for consecutive values of n, starting with n=0.

# Number spiral diagonals
## Problem 28 
Starting with the number 1 and moving to the right in a clockwise direction a 5 by 5 spiral is formed as follows:

21 22 23 24 25
20  7  8  9 10
19  6  1  2 11
18  5  4  3 12
17 16 15 14 13

It can be verified that the sum of the numbers on the diagonals is 101.

What is the sum of the numbers on the diagonals in a 1001 by 1001 spiral formed in the same way?

# Distinct powers
## Problem 29 
Consider all integer combinations of ab for 2 ≤ a ≤ 5 and 2 ≤ b ≤ 5:

22=4, 23=8, 24=16, 25=32
32=9, 33=27, 34=81, 35=243
42=16, 43=64, 44=256, 45=1024
52=25, 53=125, 54=625, 55=3125
If they are then placed in numerical order, with any repeats removed, we get the following sequence of 15 distinct terms:

4, 8, 9, 16, 25, 27, 32, 64, 81, 125, 243, 256, 625, 1024, 3125

How many distinct terms are in the sequence generated by ab for 2 ≤ a ≤ 100 and 2 ≤ b ≤ 100?

# Digit fifth powers
## Problem 30 
Surprisingly there are only three numbers that can be written as the sum of fourth powers of their digits:

1634 = 14 + 64 + 34 + 44
8208 = 84 + 24 + 04 + 84
9474 = 94 + 44 + 74 + 44
As 1 = 14 is not a sum it is not included.

The sum of these numbers is 1634 + 8208 + 9474 = 19316.

Find the sum of all the numbers that can be written as the sum of fifth powers of their digits.

# Coin sums 
## Problem 31 
In England the currency is made up of pound, £, and pence, p, and there are eight coins in general circulation:

1p, 2p, 5p, 10p, 20p, 50p, £1 (100p) and £2 (200p). It is possible to make £2 in the following way:

1×£1 + 1×50p + 2×20p + 1×5p + 1×2p + 3×1p How many different ways can £2 be made using any number of coins?

In [19]:
coins = [200,100,50,20,10,5,2,1]
target = 200
combinations = []

def coinproblem(inputset):
    fixedvalue = 0
    for n in range(0,len(inputset)):            
        fixedvalue += inputset[n] * coins[n]
        
    if len(coins)-len(inputset) == 1:
        if (target-fixedvalue)%coins[-1] == 0:
            combinations.append(inputset + [(target-fixedvalue)//coins[-1]])
    else:        
        for j in range(0,((target-fixedvalue)//coins[len(inputset)]+1)):
            coinproblem(list(inputset + [j]))

coinproblem([])
            
for i in combinations:
    fixedvalue = 0
    for j in range(0,len(coins)):        
        fixedvalue += i[j] * coins[j]
    if fixedvalue != 200:
        print('error')
        
print('There are {} possible combnations that add to £2'.format(len(combinations)))

There are 73682 possible combnations that add to £2


# Pandigital products
## Problem 32 
We shall say that an n-digit number is pandigital if it makes use of all the digits 1 to n exactly once; for example, the 5-digit number, 15234, is 1 through 5 pandigital.

The product 7254 is unusual, as the identity, 39 × 186 = 7254, containing multiplicand, multiplier, and product is 1 through 9 pandigital.

Find the sum of all products whose multiplicand/multiplier/product identity can be written as a 1 through 9 pandigital.

HINT: Some products can be obtained in more than one way so be sure to only include it once in your sum.

# Digit cancelling fractions
## Problem 33 
The fraction 49/98 is a curious fraction, as an inexperienced mathematician in attempting to simplify it may incorrectly believe that 49/98 = 4/8, which is correct, is obtained by cancelling the 9s.

We shall consider fractions like, 30/50 = 3/5, to be trivial examples.

There are exactly four non-trivial examples of this type of fraction, less than one in value, and containing two digits in the numerator and denominator.

If the product of these four fractions is given in its lowest common terms, find the value of the denominator.

# Digit factorials 
## Problem 34 
145 is a curious number, as 1! + 4! + 5! = 1 + 24 + 120 = 145.

Find the sum of all numbers which are equal to the sum of the factorial of their digits.

Note: as 1! = 1 and 2! = 2 are not sums they are not included ANS=(145,40585)

https://zero-factorial.com/whatis.html

Simple answer: 0! (read "Zero Factorial") is defined to equal 1.

Involved answer(s): There are several proofs that have been offered to support this common definition.

Example (1) If n! is defined as the product of all positive integers from 1 to n, then: 1! = 11 = 1 2! = 12 = 2 3! = 12*3 = 6 4! = 123*4 = 24 ... n! = 123*...(n-2)(n-1)n and so on. Logically, n! can also be expressed n*(n-1)! .

Therefore, at n=1, using n! = n(n-1)! 1! = 10! which simplifies to 1 = 0! Example (2) The idea of the factorial (in simple terms) is used to compute the number of permutations (combinations) of arranging a set of n numbers.

n: Number of Permutations (n!): Visual example: 1 1 {1} 2 2 {1,2}, {2,1} 3 6 {1,2,3}, {1,3,2}, {2,1,3}, {2,3,1}, {3,1,2}, {3,2,1} 10 3,628,800 ummm, you get the idea... Therefore,

0 1 { } It can be said that an empty set can only be ordered one way, so 0! = 1.

In [9]:
#Create dictionary for faster processing
factorials = {'0':1} 
output = 1
for i in range(1,10):
    output *= i
    factorials.update({str(i):output})
    
print(factorials)

{'0': 1, '1': 1, '2': 2, '3': 6, '4': 24, '5': 120, '6': 720, '7': 5040, '8': 40320, '9': 362880}


In [11]:
def factorial(num):
    if num == 0: return 1  #I dont really get this but it is true...  
    output = 1 
    for i in range(1, num+1): 
        output*=i
    return(output)

def digit_factorial_sum(num):
    output = 0
    while num > 0:
        output += factorials[str(num%10)]
        num = num//10    
    return output      


In [12]:
limit = 50000

digit_factorials = []
for n in range(3,limit):
    if n == digit_factorial_sum(n):
        digit_factorials.append(n)
    
print('There are {} digit factorials below {}. The largest is {}. The sum of these is {}'.format(len(digit_factorials),limit,digit_factorials[-1],sum(digit_factorials)))  
print(','.join(map(str,digit_factorials)))   

There are 2 digit factorials below 50000. The largest is 40585. The sum of these is 40730
145,40585


# Circular primes 
## Problem 35 
The number, 197, is called a circular prime because all rotations of the digits: 197, 971, and 719, are themselves prime.

There are thirteen such primes below 100: 2, 3, 5, 7, 11, 13, 17, 31, 37, 71, 73, 79, and 97.

How many circular primes are there below one million?

In [21]:
#generates all rotations of digits in an integer
def num_rotates(num):
    output = []
    for i in range(0,len(str(num))):
        output.append(int(str(num)[i:]+str(num)[:i]))
    return output

limit = 10**6
circular = []
for n in [n for n in primes if n < limit]:
    circularbool = None       
    for p in num_rotates(n):
        if not isprime(p):
            circularbool = False
            break
    if circularbool != False:
        circular.append(n)
print('There are {} circular primes below {}. The largest is {}.'.format(len(circular),limit,circular[-1]))  
print(','.join(map(str,circular)))

There are 55 circular primes below 1000000. The largest is 999331.
2,3,5,7,11,13,17,31,37,71,73,79,97,113,131,197,199,311,337,373,719,733,919,971,991,1193,1931,3119,3779,7793,7937,9311,9377,11939,19391,19937,37199,39119,71993,91193,93719,93911,99371,193939,199933,319993,331999,391939,393919,919393,933199,939193,939391,993319,999331


# Double-base palindromes 
# Problem 36 
The decimal number, 585 = 10010010012 (binary), is palindromic in both bases.

Find the sum of all numbers, less than one million, which are palindromic in base 10 and base 2.

(Please note that the palindromic number, in either base, may not include leading zeros.)m

doubleBasePalindromes(1000) should return 1772. doubleBasePalindromes(50000) should return 105795. doubleBasePalindromes(500000) should return 286602. doubleBasePalindromes(1000000) should return 872187.

In [13]:
limit = 10**6
double_palindromes = []
for n in range(1,limit):
    #print(n, str(n),str(n)[::-1],"{0:b}".format(n),"{0:b}".format(n)[::-1])
    if str(n) == str(n)[::-1] and "{0:b}".format(n) == "{0:b}".format(n)[::-1]:
        double_palindromes.append(n)
    
print('There are {} double palindromes below {}. The sum of these is {}.'.format(len(double_palindromes),limit,sum(double_palindromes)))


There are 19 double palindromes below 1000000. The sum of these is 872187.


# Truncatable primes 
## Problem 37 
The number 3797 has an interesting property. Being prime itself, it is possible to continuously remove digits from left to right, and remain prime at each stage: 3797, 797, 97, and 7. Similarly we can work from right to left: 3797, 379, 37, and 3.

Find the sum of the only eleven primes that are both truncatable from left to right and right to left.

NOTE: 2, 3, 5, and 7 are not considered to be truncatable primes.

In [22]:
def istrunctableprime(num):
    numstr = str(num)
    for i in range(0,len(numstr)):
        if isnotprime(int(numstr[i:])) : return False
        if isnotprime(int(numstr[:i+1])) : return False                        
    return True        

istrunctableprime(3797)

limit = 10**6
trunctable = []
for n in [n for n in primes if (n > 9) and (n < limit)]:
    if istrunctableprime(n):
        trunctable.append(n)

print('There are {} trunctable primes below {}. The largest is {}.'.format(len(trunctable),limit,trunctable[-1]))  
print(','.join(map(str,trunctable)))


sum(trunctable)

NameError: name 'primess' is not defined

# Consecutive prime sum
## Problem 50 
The prime 41, can be written as the sum of six consecutive primes:

41 = 2 + 3 + 5 + 7 + 11 + 13
This is the longest sum of consecutive primes that adds to a prime below one-hundred.

The longest sum of consecutive primes below one-thousand that adds to a prime, contains 21 terms, and is equal to 953.

Which prime, below one-million, can be written as the sum of the most consecutive primes?

In [5]:
#loadprimes - DO THIS IF YOU DO NOT HAVE TWO HOURS
inputfile = open('C:\\Users\\alexp\\primes.csv','r')
primes = list(map(int,inputfile.read().split(',')))
inputfile.close()

In [7]:
target = 100 

consecutive_primes = []
for i in range(0,len([i for i in primes if i < target])):
    n = 1
    while sum(primes[i:n+1])<target:
        n+=1
        if sum(primes[i:n]) in primes:
            consecutive_primes.append((sum(primes[i:n]),primes[i:n]))
  #  print('{} {} {} {}'.format(i, primes[i], n, sum(primes[i:n])))


In [8]:
consecutive_primes

[(5, [2, 3]),
 (17, [2, 3, 5, 7]),
 (41, [2, 3, 5, 7, 11, 13]),
 (3, [3]),
 (5, [5]),
 (23, [5, 7, 11]),
 (53, [5, 7, 11, 13, 17]),
 (7, [7]),
 (31, [7, 11, 13]),
 (67, [7, 11, 13, 17, 19]),
 (11, [11]),
 (41, [11, 13, 17]),
 (83, [11, 13, 17, 19, 23]),
 (13, [13]),
 (17, [17]),
 (59, [17, 19, 23]),
 (19, [19]),
 (71, [19, 23, 29]),
 (23, [23]),
 (83, [23, 29, 31]),
 (29, [29]),
 (97, [29, 31, 37]),
 (31, [31]),
 (37, [37]),
 (41, [41]),
 (43, [43]),
 (47, [47]),
 (53, [53]),
 (59, [59]),
 (61, [61]),
 (67, [67]),
 (71, [71]),
 (73, [73]),
 (79, [79]),
 (83, [83]),
 (89, [89]),
 (97, [97])]

# Maximum path sum II
## Problem 67 
By starting at the top of the triangle below and moving to adjacent numbers on the row below, the maximum total from top to bottom is 23.

3
7 4
2 4 6
8 5 9 3

That is, 3 + 7 + 4 + 9 = 23.

Find the maximum total from top to bottom in triangle.txt (right click and 'Save Link/Target As...'), a 15K text file containing a triangle with one-hundred rows.

NOTE: This is a much more difficult version of Problem 18. It is not possible to try every route to solve this problem, as there are 299 altogether! If you could check one trillion (1012) routes every second it would take over twenty billion years to check them all. There is an efficient algorithm to solve it. ;o)

In [11]:
inputfile = open('C:\\Users\\alexp\\Downloads\\Data Science\\Project Euler\\p067_triangle.txt','r')
triangle = [list(map(int,row.split())) for row in inputfile.read().split('\n')]
inputfile.close()
triangle = [i for i in triangle if len(i) > 0]

# (value, maxsum, right)

for i in range(len(triangle)-1,-1,-1):
    for j in range(len(triangle[i])):
        if i == len(triangle)-1:
            triangle[i][j] = (triangle[i][j],triangle[i][j],)
        else:
            if triangle[i+1][j][1] > triangle[i+1][j+1][1]:
                triangle[i][j] = (triangle[i][j],triangle[i][j]+triangle[i+1][j][1],False)
            else:
                triangle[i][j] = (triangle[i][j],triangle[i][j]+triangle[i+1][j+1][1],True)                   
        
print('The highest path with a value of {} is :'.format(triangle[0][0][1]))
print()

r = 0
bestpath = []
for i in range(len(triangle)-1):
    print(triangle[i][r][0], end='\t')
    if triangle[i][r][2]:
        r += 1
print()

The highest path with a value of 7273 is :

59	73	52	53	87	57	92	81	81	79	81	32	86	82	97	55	97	36	62	65	90	93	95	54	71	77	68	71	94	8	89	54	42	90	84	91	31	71	93	94	53	69	73	99	89	47	80	96	81	52	98	38	91	78	90	70	61	17	11	75	74	55	81	87	89	99	73	88	95	68	37	87	73	77	60	82	87	64	96	65	47	94	85	51	87	65	65	66	91	83	72	24	98	89	53	82	57	99	98	
